### PR TITLE
Add support for highlighting with CSS classes

### PR DESCRIPTION
**Pull request by camdencheek**
This PR adds a query parameter `css`, which toggles rendering the HTML as a table with CSS classes. 

A couple of notes:
- This moves table generation logic into this service, removing the need to parse the HTML and rebuild it in the format we need
- We copy the `Escape` struct from `syntect` because it is not exported. I've opened a PR to export it [here](https://github.com/trishume/syntect/pull/330), but I'd like to not block on that

Related PRs: sourcegraph/gosyntect#13, sourcegraph/sourcegraph#19986

### CODE REVIEW

File: src/main.rs
Comment (camdencheek) [inline src/main.rs]: I implemented this as a completely separate endpoint because this version returns results in a different format (table vs raw spans). Additionally, it let us remove some deprecated fields from the query. 

However, it does lead to a bit of duplicate code. I plan to clean that up as much as possible before merging.

File: src/classed_table_generator.rs
Comment (camdencheek) [inline src/classed_table_generator.rs]: Note that this implementation looks quite long, but most of it is quite simple, and will replace ~200 lines of very complex go code that currently parses the results as HTML and reforms it into this table format.

File: Cargo.toml
Comment (camdencheek) [inline Cargo.toml]: Currently using my fork because of slimsag/syntect#8

File: Dockerfile
Comment (camdencheek) [inline Dockerfile]: that version of git no longer exists in the package repos. Was there a reason that specific version was pinned?

Comment (emidoots) [reply Dockerfile]: It's good practice to pin dependencies in general, Hadolint enforces this in our main repo but we haven't enabled it here yet. Can you pin the version to whatever does exist?

File: src/main.rs
Comment (emidoots) [reply src/main.rs]: sounds good

File: src/classed_table_generator.rs
Comment (emidoots) [reply src/classed_table_generator.rs]: One reason I chose to write this logic in Go originally was because we don't have anyone else at Sourcegraph who is able to dive into this code-base and work on it if we need to. My time these days is also particularly limited, so if there are changes needed to this code in the future it will likely be a "camden or.. we may not be able to do it any time soon"

I don't think that is a blocker here, but thought it worth raising as a point.

Also, let's make sure to add tests for this code before merging.

**Comment by emidoots** [comment]
Once the upstream PR is merged, we should land a PR (separate from this one) to update slimsag/syntect to the latest revision. This will require:

* Merging `trishume/syntect@master` -> `slimsag/syntect@master`
* Re-running `make testdata` (in the syntect repo) to bundle in our syntax files in.
* Running `cargo update -p syntect` in this repository, and regenerating https://github.com/sourcegraph/sourcegraph/blob/main/third-party-licenses/ThirdPartySyntectServer.dependencies.json using the `cargo-license` tool (I have promised Dan I would enforce we keep this up-to-date or reject changes until we do, this is very important)
* Publishing a new Docker image, deploying it on sourcegraph.com, and monitoring to make sure there aren't any major regressions

**Comment by camdencheek** [comment]
Great -- I'll put this on hold until the upstream change is merged. It just got approved, so it hopefully shouldn't take very long (🤞 ). 

Other than the steps you outlined after pulling in the upstream, I think this PR should be ready for review. I'll leave it as a draft for the time being since it's not in a merge-ready state, but I don't expect it to change much from here. I added a few unit tests to exercise the table generator a bit, but I'll also make sure the integration tests are updated in `sourcegraph/sourcegraph`.

**Comment by camdencheek** [comment]
Given there hasn't been movement upstream, I plan to just copy that struct into `syntect_server` with a TODO to replace it with the upstream version once it's merged. This way, there's no `slimsag/syntect` code change, and I can unblock this.

File: Cargo.lock
Comment (emidoots) [inline Cargo.lock]: I wouldn't expect this file to change unless the version of Rust is changing too (note that this may introduce other issues in compiling the Docker container.)

Can you either undo these changes, or upgrade the version of Rust in [this file](https://github.com/sourcegraph/syntect_server/blob/master/rust-toolchain) and the Dockerfile?

File: src/main.rs
Comment (emidoots) [inline src/main.rs]: Why should these tests be in the main file and not in the `src/classed_table_generator.rs` file? This seems odd

Comment (camdencheek) [reply src/main.rs]: Because they are testing the `css_table_highlight` function, which lives in `main.rs`. I can move that to the other file along with the tests -- probably better organized there anyways.

File: src/css_table.rs
Comment (emidoots) [inline src/css_table.rs:10]:
```suggestion
```

File: src/css_table.rs
Comment (emidoots) [inline src/css_table.rs]:
```suggestion
```

File: src/css_table.rs
Comment (emidoots) [inline src/css_table.rs]:
```suggestion
```

File: src/main.rs
Comment (emidoots) [inline src/main.rs]:
```suggestion

#[post("/css_table", format = "application/json", data = "<q>")]
```

File: src/css_table.rs
Comment (camdencheek) [reply src/css_table.rs:10]: Ran it through `rustfmt`, which seemed to fix all these

**Comment by emidoots** [comment]
@camdencheek I want to talk a bit about the decision to introduce a new `/css_table` endpoint vs. just adding an option to the existing `/` endpoint. From my perspective, this only adds complexity:

* It could be easily represented with just a `css` boolean flag on the `/` index route.
* Introducing the new route adds a non-trivial amount of duplication in gosyntect: https://github.com/sourcegraph/gosyntect/pull/13/files
* It's not clear to me what the plan is (or if there is one) to reduce this complexity and get us back to a single route over time.

Thoughts?

**Comment by camdencheek** [comment]
For sure -- I'd be happy to reduce it down to one route. Mostly, I didn't want to break the existing routes. There is non-trivial change of behavior here, so it's not quite as simple as add a flag.

Some of the differences:
1) We return a table instead of spans. This means the client has to process the response differently depending on whether or not `css` is enabled.
2) Theme is only a valid argument for non-CSS
3) MaxLineLength is only usable with the new table generator

Some potential solutions:
1. This one is more difficult, and the main reason I avoided combining the endpoints. I think it would be pretty surprising for an endpoint to return two completely divergent result formats. We could add support for highlighting from theme files in the table format, but that felt like a separate effort. Maybe we have two toggles: `css` and `table`, and just return an error if a caller asks for a combination we don't support?
2. This can be documented as "not used if CSS is set"
3. We were never allowing configuration of MaxLineLength before (just setting it to 2000), so we could pretty simply just not expose this option


The other path we could take here is just not expect `syntect_server` and `gosyntect` to be backwards compatible. I don't know if these are used by any projects outside of Sourcegraph -- you'd be a better one to make that call. (FWIW, a [quick search](https://sourcegraph.com/search?q=context:global+gosyntect+select:repo&patternType=literal) only shows a result for a sourcegraph fork)

**Comment by emidoots** [comment]
> We return a table instead of spans. This means the client has to process the response differently depending on whether or not css is enabled.

I think this is fine, we can just document that if the `css` field is true then it changes the response.

> Theme is only a valid argument for non-CSS

Same thing here, just document it is ignored if `css=true`

> MaxLineLength is only usable with the new table generator

Also seems fine with documentation.

**Comment by camdencheek** [comment]
Sounds good -- I'll merge the endpoints 👍

**Comment by emidoots** [comment]
> This one is more difficult, and the main reason I avoided combining the endpoints. I think it would be pretty surprising for an endpoint to return two completely divergent result formats. We could add support for highlighting from theme files in the table format, but that felt like a separate effort. Maybe we have two toggles: css and table, and just return an error if a caller asks for a combination we don't support?

I don't think we need to support this. It would be good to have it return the same data structure eventually, but for now I don't see any major issue in returning separate data structures.

In the future (ideal world), we'd add a new `table: true` field and have it return the same data structure always, and document that `table:false` is deprecated and eventually remove it.

**Comment by emidoots** [comment]
> The other path we could take here is just not expect syntect_server and gosyntect to be backwards compatible. I don't know if these are used by any projects outside of Sourcegraph -- you'd be a better one to make that call. (FWIW, a quick search only shows a result for a sourcegraph fork)

I think syntect_server _may_ be used outside of Sourcegraph by like 1-2 people, but if it is I'd expect the users to follow pretty closely and/or only update when it makes sense. This is clearly a Sourcegraph-focused project today, and backwards compatibility with the API is not a huge concern/deal. I (and any other users) would rather us arrive at a good, solid API than worry about backwards compat.

I don't think gosyntect is used by anybody except us.

**Comment by camdencheek** [comment]
Preliminary commit just pushed. Working on the gosyntect changes now so I can test it more thoroughly

**Comment by camdencheek** [comment]
@slimsag Updated it to use the same endpoint. It made the `gosyntect` PR _significantly_ simpler. Thanks for the suggestion.

File: src/main.rs
Comment (emidoots) [inline src/main.rs:112]: Hmmm, this `or_else` logic _sounds_ smart at first glance but I'm actually not so sure about it. I think this could well make it harder in practice to figure out why a language is being highlighted a certain way in the case of the wrong language being picked up.

Let's try it and see, but maybe something to revisit later.

**Review by emidoots — APPROVED** [review]
Looks great, thanks for taking this to the finish line and being the first person in over 4 years(!!!) to work on syntect_server besides me :)

File: src/main.rs
Comment (camdencheek) [reply src/main.rs:112]: Fair enough. Definitely not attached to it

**Comment by camdencheek** [comment]
Great! thanks for all your help and patience 😄

**Pull request merged**
